typhos/cache.py

```diff
@@ -341,3 +386,291 @@ def add_path(self, path):
+    # should be bumped (or the db should be wiped entirely)
+    _version = 0
+    _table_name = f'describe_cache_v{_version}'
+
+    def __init__(self, path=None):
+        super().__init__()
+        self.log = logging.getLogger(__name__ + '._DescribeDatabase')
+        self._describe_cache = None
+        self._preload_cache = {}
+        self._init_queries()
+        self._init_database(path or get_describe_database_path())
+
+    @property
+    def path(self):
+        """The database file path (or ':memory:')."""
+        return self._path
+
+    @property
+    def describe_cache(self):
+        """The in-memory describe cache - _GlobalDescribeCache."""
+        return self._describe_cache
+
+    @describe_cache.setter
+    def describe_cache(self, cache):
+        if self._describe_cache is not None:
+            raise ValueError('describe_cache already set')
+
+        self._describe_cache = cache
+        cache.new_description.connect(self._new_description)
+        # NOTE: this does not set the describe cache's persistent cache.
+
+    def _init_queries(self):
+        """Pre-create all queries sent to the database connection."""
+        cols = ', '.join(f'{key} {self._type_map[value]}'
+                         for key, value in self._columns.items()
+                         if value is not None)
+
+        self._create_query = f'''
+            CREATE TABLE IF NOT EXISTS "{self._table_name}" (
+                name text primary key,
+                class text not null,
+                pvname text,
+                setpoint_pvname text,
+                {cols})
+        '''
+
+        col_names = ', '.join(self._columns)
+        empty_names = ', '.join('?' * len(self._columns))
+        self._insert_query = f'''
+            REPLACE INTO "{self._table_name}"
+            ({col_names})
+            VALUES ({empty_names})
+        '''
+
+        self._select_query = f'''
+            SELECT * FROM "{self._table_name}" WHERE
+            name=? AND
+            class=? AND
+            pvname=? AND
+            setpoint_pvname=?
+        '''
+
+        self._select_key_query = f'''
+            SELECT name, class, pvname, setpoint_pvname
+            FROM "{self._table_name}"
+        '''
+
+        self._select_count_query = f'SELECT COUNT(*) FROM {self._table_name}'
+
+        self.log.debug('Create query: %s', self._create_query)
+        self.log.debug('Insert query: %s', self._insert_query)
+        self.log.debug('Select query: %s', self._select_query)
+
+    def _init_database(self, path):
+        self._path = path
+        self._con = sqlite3.connect(path)
+        # Support __getitem__ in row access:
+        self._con.row_factory = sqlite3.Row
+
+        self.log.info('Describe database path: %s', path)
+        self._con.execute(self._create_query)
+
+        t0 = time.time()
+        self._preload()
+        elapsed = time.time() - t0
+        self.log.info('Took %d ms to preload the cache', int(elapsed * 1000))
```

Comment:
```suggestion
        self.log.debug('Took %d ms to preload the cache', int(elapsed * 1000))
```
This isn't useful information for the end-user so it belongs in debug:
```
[2020-05-07 14:42:13] - INFO - Loading "pcdsdevices.device_types.PPM[{'prefix':'IM3L0:PPM'}]" ...
[2020-05-07 14:42:15] - INFO - Loading Tools ...
[2020-05-07 14:42:15] - INFO - Describe database path: /reg/neh/home4/zlentz/.cache/typhos/typhos_describe.sqlite
[2020-05-07 14:42:15] - INFO - Took 2 ms to preload the cache
[2020-05-07 14:42:16] - INFO - Adding devices ...
[2020-05-07 14:42:42] - INFO - Launching application ...
[2020-05-07 14:44:02] - INFO - Execution complete!
```
```
[2020-05-07 14:44:20] - INFO - Loading "pcdsdevices.device_types.PPM[{'prefix':'IM3L0:PPM'}]" ...
[2020-05-07 14:44:21] - INFO - Loading Tools ...
[2020-05-07 14:44:22] - INFO - Describe database path: /reg/neh/home4/zlentz/.cache/typhos/typhos_describe.sqlite
[2020-05-07 14:44:28] - INFO - Took 6102 ms to preload the cache
[2020-05-07 14:44:28] - INFO - Adding devices ...
[2020-05-07 14:46:01] - INFO - Launching application ...
[2020-05-07 14:46:17] - INFO - Execution complete!
```

Reply:
Agreed and accepted.

However, that 6 seconds is atrocious. Perhaps that indicates this just shouldn't be merged at all (or at minimum an alternative pre-loading method should be done)